src/fw_update.c

```diff
@@ -75,26 +78,60 @@ static enum golioth_status fw_write_block_cb(const struct golioth_ota_component
     return status;
 }
 
-static enum golioth_status golioth_fw_update_report_state_sync(struct golioth_client *client,
-                                                               enum golioth_ota_state state,
-                                                               enum golioth_ota_reason reason,
-                                                               const char *package,
-                                                               const char *current_version,
-                                                               const char *target_version,
-                                                               int32_t timeout_s)
+enum golioth_status golioth_fw_update_report_state_sync(struct fw_update_component_context *ctx,
+                                                        enum golioth_ota_state state,
+                                                        enum golioth_ota_reason reason,
+                                                        bool report_component_name,
+                                                        bool report_current_version,
+                                                        bool report_target_version)
 {
+    if (!ctx)
+    {
+        return GOLIOTH_ERR_NULL;
+    }
+
     if (_state_callback)
     {
         _state_callback(state, reason, _state_callback_arg);
     }
 
-    return golioth_ota_report_state_sync(client,
-                                         state,
-                                         reason,
-                                         package,
-                                         current_version,
-                                         target_version,
-                                         timeout_s);
+    enum golioth_status status;
+    uint32_t backoff_s = FW_REPORT_RETRIES_INITAL_DELAY_S;
+    int retries_remaining = FW_REPORT_MAX_RETRIES;
+
+retry_report:
```

Comment:
Nit: it looks that replacing label+goto should be doable with `while (retries_remaining--)` logic. Behavior should be the same, but it will  probably be a bit easier to read due to widely used while/for loops compared to goto (with the main exception being error handling).

Reply:
Thanks for noticing. This is much better as a while loop.